common/quotas/priority_rate_limiter_impl.go

```diff
@@ -0,0 +1,163 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+)
+
+type (
+	// PriorityRateLimiterImpl is a wrapper around the golang rate limiter
+	PriorityRateLimiterImpl struct {
+		apiToPriority          map[string]int
+		priorityToRateLimiters map[int]RateLimiter
+
+		// priority value 0 means highest priority
+		// sorted rate limiter from lot priority value to high priority value
+		priorityToIndex map[int]int
+		rateLimiters    []RateLimiter
+	}
+)
+
+// NewPriorityRateLimiter returns a new rate limiter that can handle dynamic
+// configuration updates
+func NewPriorityRateLimiter(
+	apiToPriority map[string]int,
```

Comment:
do we need to cross-check validation between the priority value in the apiToPriority map and make sure a corresponding key exists in priorityToRateLimiters map?

Reply:
added an input validation